src/CryptState.h

```diff
@@ -32,23 +28,55 @@ class CryptState {
 		unsigned int uiRemoteLost;
 		unsigned int uiRemoteResync;
 
-		AES_KEY	encrypt_key;
-		AES_KEY decrypt_key;
 		Timer tLastGood;
 		Timer tLastRequest;
 		bool bInit;
 		CryptState();
+		virtual ~CryptState() {};
+
+		virtual bool isValid() const = 0;
+		virtual void genKey() = 0;
+		virtual void setKey(const std::string& rkey, const std::string& eiv, const std::string& div) = 0;
+		virtual void setRawKey(const std::string& rkey) = 0;
+		virtual void setEncryptIV(const std::string& iv) = 0;
+		virtual void setDecryptIV(const std::string& iv) = 0;
+		virtual std::string getRawKey() = 0;
+		virtual std::string getEncryptIV() = 0;
+		virtual std::string getDecryptIV() = 0;
```

Comment:
Using `std::string` for a key (which is a raw byte vector) invites null-termination bugs.
Better use `using RawByteVec = std::vector<unsigned char>` (or more explicit `std::vector<std::byte>` if you do not want to do interger arithmetic).

Reply:
Good catch!

Since our baseline is currently C++14, I would use `std::vector<uint8_t>`.

Reply:
Thanks for bringing that up! I don't think null-termination will become a problem since the length of the string is stored inside std::string and when initializing a std::string, I specified a length. Most importantly, this is how protobuf does. (I just learned this from protobuf ^^)

Reply:
I also don't see a problem in using `std::string` for this purpose...